_posts/2020-02-05-keycloak-z-angularem-i-keycloakiem.md

```diff
@@ -0,0 +1,427 @@
+
+![Keycloak login](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/keycloak_login.png)
+
+Jak widzimy, adres zawiera informacje, takie jak nazwa protokołu, realm, client czy adres, na który mamy zostać przekierowani po zalogowaniu.
+
+Jeśli zalogujemy się użytkownikiem posiadającym rolę `user_role` (`user:password`), to zostaniemy przekierowani z powrotem do naszej aplikacji:
+
+![Public component](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/public_component_logged.png)
+
+Dodatkowo mamy jeszcze tylko przycisk `Logout (frontend)`. Jesteśmy teraz zalogowani, więc możemy przejść do `Protected`:
+
+![Protected component](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/protected_component.png)
+
+Zadziała również `Backend Hello`:
+
+![Backend Hello](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/backend_hello.png)
+
+Jeśli wylogujemy się teraz z frontendu - `Logout (frontend)`, to zniknie przycisk `Logout (frontend)` i nie będziemy mieli już dostępu do `Protected`.  
+Cały czas jednak będziemy mieli dostęp do `Backend Hello`. Dopiero kiedy wylogujemy się z backendu - `Logout (backend)`, to stracimy dostęp do `Backend Hello`.  
+Jednak jeśli będąc zalogowanym użytkownikiem, wylogujemy się z backendu a nie frontendu, to automatycznie zostaniemy również wylogowani z frontendu.
+
+Dlaczego?
+
+O tym jak działa Keycloak i tokeny którymi się posługujemy (a dokładniej standard OpenID Connect) nie jest tematem tego wpisu. Co do kwestii wylogowywania, to temat (ze względu na złożoność implementacji) nadaje się na osobny wpis, szczególnie gdy mówimy o rozproszonych systemach z wieloma instancjami aplikacji, stojącymi za loadbalancerem.
+
+Nasz projekt demo posiada bardzo prosty mechanizm wylogowywania.
+
+Wylogowanie z frontendu wywoła metodę `logout` na serwisie biblioteki `keycloak-angular`:
+
+```typescript
+logout = async (): Promise<void> => await this.keycloakService.logout();
+```
+
+Natomiast wylogowanie z backendu uderzy na `/api/logout`, co spowoduje wykonanie na backendzie:
+
+```java
+fun logout(request: HttpServletRequest, response: HttpServletResponse) {
+    request.logout()
+    response.sendRedirect("/")
+}
+```
+
+Na koniec sprawdźmy jeszcze użytkownika bez roli `user_role` (`user2:password2`).
+
+Po zalogowaniu, w aplikacji powinniśmy zobaczyć te same przyciski w menu, co na użytkowniku posiadającym wymaganą rolę.  
+Jest tak, ponieważ ukrywanie przycisków uzależniliśmy tylko od tego, czy użytkownik jest zalogowany, a nie czy ma uprawnienia do routa:
+
+```typescript
+ngOnInit(): void {
+    this.keycloakService.isLoggedIn().then(isLogged => this.isLogged = isLogged);
+}
+```
+
+Uderzenie na `Backend Hello` również nie zadziała.
+
+Uwierzytelnianie możemy bardzo szybko wyłączyć w obu aplikacjach, ustawiając `keycloak.enabled` w `application.yml` na `false`.  
+Może być to przydatne np. na środowiskach testowych.
+
+## Podsumowanie
```

Comment:
Ten punkt to nie jest podsumowanie, piszesz tutaj o alternatywnych rozwiązaniach i dalszej rozbudowie. Zmieniłbym tytuł i dopisał właściwe podsumowanie ze sztampowym "Mam nadzieję, że udało mi się przybliżyć tamatykę integracji Keylocka z aplikacją opratą o Spring Boot i Angular" i rzucił garść linków " for further readings"